File: ZAJECIA_1/ZAD7_8/Ciagi_graficzne.java
package JAVA;
import java.util.*;

public class Ciagi_graficzne
{
    //Metoda Havla-Hakimiego ze zlozonoscia O(n²)
    public static boolean havlaHakimi(List<Integer> stopnie)
    {
        while(true)
        {
            //Sortujemy malejaco
            Collections.sort(stopnie, Collections.reverseOrder());
            //Jesli wszystkie stopnie sa rowne 0 to graf istnieje
            if(stopnie.get(0) == 0)
            {
                return true;
            }
            int najwyzszy = stopnie.get(0);
            stopnie.remove(0);
            //Warunek gdy stopien jest za duzy
            if(najwyzszy > stopnie.size())
            {
                return false;
            }
            //Zmniejszanie stopni wierzcholkow
            for(int i = 0; i < najwyzszy; i++)
            {
                stopnie.set(i, stopnie.get(i) - 1);
                if(stopnie.get(i) < 0)
                {
                    return false;
                }
            }
        }
    }

    //Metoda Erdosa-Gallai ze zlozonoscia O(n)
    public static boolean erdosGallai(List<Integer> stopnie)
    {
        //Sortujemy malejaco
        Collections.sort(stopnie, Collections.reverseOrder());
        int suma_stopni = stopnie.stream().mapToInt(Integer::intValue).sum();
        //Suma stopni musi byc parzysta
        if(suma_stopni % 2 != 0)
        {
            return false;
        }
        int sumaLewa = 0;
        int sumaPrawa;
        int n = stopnie.size();
        for(int k = 1; k <= n; k++)
        {
            sumaLewa = sumaLewa + stopnie.get(k - 1);
            sumaPrawa = k * (k - 1);
            for(int i = k; i < n; i++)
            {
                sumaPrawa = sumaPrawa + Math.min(k, stopnie.get(i));
            }
            if(sumaLewa > sumaPrawa)
            {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args)
    {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Podaj liczbe wierzcholkow: ");
        int liczbaWierzcholkow = scanner.nextInt();
        List<Integer> stopnie = new ArrayList<>();
        for(int i = 0; i < liczbaWierzcholkow; i++)
        {
            stopnie.add(scanner.nextInt());
        }
        System.out.println("\nWynik metody Havla-Hakimiego: " + (havlaHakimi(new ArrayList<>(stopnie)) ? "CIAG JEST GRAFICZNY" : "CIAG NIE JEST GRAFICZNY"));
        System.out.println("Wynik metody Erdosa-Gallai: " + (erdosGallai(new ArrayList<>(stopnie)) ? "CIAG JEST GRAFICZNY" : "CIAG NIE JEST GRAFICZNY"));
        scanner.close();
    }
}

File: ZAJECIA_1/ZAD7_8/Ciagi_graficzne.py
"""Metoda Havla-Hakimiego ze zlozonoscia O(n²)"""
def havla_hakimi(stopnie):
    while True:
        #Sortowanie malejaco
        stopnie.sort(reverse=True)
        #Jesli wszystkie stopnie sa rowne 0, to graf istnieje
        if(stopnie[0] == 0):
            return True
        #Pobieramy najwiekszy stopien
        najwyzszy = stopnie.pop(0)
        #Warunek czy stopien nie jest zbytnio duzy
        if(najwyzszy > len(stopnie)):
            return False
        #Odejmujemy 1 od kolejnych elementow
        for i in range(najwyzszy):
            stopnie[i] = stopnie[i] - 1
            if(stopnie[i] < 0):
                return False

"""Metoda Erdosa-Gallai ze zlozonoscia O(n)"""
def erdos_gallai(stopnie):
    #Sortujemy malejaco
    stopnie.sort(reverse=True)
    suma_stopni = sum(stopnie)
    #Sprawdzenie parzystosci
    if(suma_stopni % 2 != 0):
        return False
    suma_lewa = 0
    n = len(stopnie)

    for k in range(1, n + 1):
        suma_lewa = suma_lewa + stopnie[k - 1]
        suma_prawa = k * (k - 1)
        for i in range(k, n):
            suma_prawa = suma_prawa + min(k, stopnie[i])

        #Jesli warunek nie zachodzi
        if(suma_lewa > suma_prawa):
            return False
    return True
    
liczba_wierzcholkow = int(input("Podaj liczbe wierzcholkow: "))
stopnie = list(map(int, input("Podaj stopnie wierzcholkow: ").split()))

print("\nWynik metody Havla-Hakimiego:", "CIAG JEST GRAFICZNY" if havla_hakimi(stopnie[:]) else "CIAG NIE JEST GRAFICZNY")
print("Wynik metody Erdosa-Gallai:", "CIAG JEST GRAFICZNY" if erdos_gallai(stopnie[:]) else "CIAG NIE JEST GRAFICZNY")